File: kiko-api/src/repositories/qualityUsersRepository.ts
import prisma, { withRetry } from '../db/prisma.js';
import { get, set } from '../cache/cacheClient.js';
import { Decimal } from 'decimal.js';

export interface QualityFarcasterUser {
  fid: number;
  username?: string;
  displayName?: string;
  pfp?: string;
  bio?: string;
  verifications?: Array<{ address: string; protocol: string }>;
  url?: string;            // USER_DATA_TYPE_URL — personal website
  banner?: string;         // USER_DATA_TYPE_BANNER — profile banner image
  primaryAddress?: string; // USER_DATA_PRIMARY_ADDRESS_ETHEREUM
  location?: string;       // USER_DATA_TYPE_LOCATION — geo:lat,lng
  followers?: number;
  following?: number;
  totalCasts?: number;
  engagementRate?: number;
  source?: string;
  isActive?: boolean;
  hasCreatorCoin?: boolean;
  creatorCoinAddress?: string;
  lastCoinCheck?: Date;
}

const CACHE_KEY = 'farcaster:quality_users';
const CACHE_TTL = 3600; // 1 hour

export async function getQualityFids(): Promise<number[]> {
  try {
    const cached = await get(CACHE_KEY);
    if (cached) {
      return JSON.parse(cached) as number[];
    }

    const result = await prisma.qualityFarcasterUser.findMany({
      where: { isActive: true },
      orderBy: { followers: 'desc' },
      take: 2000,
      select: { fid: true }
    });

    const fids = result.map(row => row.fid);

    if (fids.length > 0) {
      await set(CACHE_KEY, JSON.stringify(fids), CACHE_TTL);
    }

    return fids;
  } catch (error) {
    console.error('[QualityUsersRepo] Error getting quality FIDs:', error);
    return [];
  }
}

export async function getQualityUsers(limit: number = 100): Promise<QualityFarcasterUser[]> {
  try {
    const result = await prisma.qualityFarcasterUser.findMany({
      where: { isActive: true },
      orderBy: { followers: 'desc' },
      take: limit
    });

    return result.map((row) => ({
      fid: row.fid,
      username: row.username || undefined,
      displayName: row.displayName || undefined,
      pfp: row.pfp || undefined,
      bio: row.bio || undefined,
      verifications: Array.isArray(row.verifications) ? row.verifications as Array<{ address: string; protocol: string }> : [],
      url: (row as any).url || undefined,
      banner: (row as any).banner || undefined,
      primaryAddress: (row as any).primaryAddress || undefined,
      location: (row as any).location || undefined,
      followers: row.followers,
      following: row.following,
      totalCasts: row.totalCasts,
      engagementRate: Number(row.engagementRate),
      source: row.source || undefined,
      isActive: row.isActive,
      hasCreatorCoin: row.hasCreatorCoin,
      creatorCoinAddress: row.creatorCoinAddress || undefined,
      lastCoinCheck: row.lastCoinCheck || undefined,
    }));
  } catch (error) {
    console.error('[QualityUsersRepo] Error getting quality users:', error);
    return [];
  }
}

export async function saveQualityUsers(users: QualityFarcasterUser[]): Promise<number> {
  if (users.length === 0) return 0;
  let savedCount = 0;

  try {
    await withRetry(async () => {
      await prisma.$transaction(async (tx) => {
        for (const user of users) {
          await tx.qualityFarcasterUser.upsert({
            where: { fid: user.fid },
            update: {
              username: user.username || undefined,
              displayName: user.displayName || undefined,
              pfp: user.pfp || undefined,
              bio: user.bio || undefined,
              verifications: user.verifications || undefined,
              followers: user.followers || undefined,
              following: user.following || undefined,
              totalCasts: user.totalCasts || undefined,
              engagementRate: user.engagementRate !== undefined ? new Decimal(user.engagementRate) : undefined,
              source: user.source || 'dune',
              isActive: true,
              lastVerifiedAt: new Date(),
              updatedAt: new Date(),
            },
            create: {
              fid: user.fid,
              username: user.username || null,
              displayName: user.displayName || null,
              pfp: user.pfp || null,
              bio: user.bio || null,
              verifications: user.verifications || [],
              followers: user.followers || 0,
              following: user.following || 0,
              totalCasts: user.totalCasts || 0,
              engagementRate: new Decimal(user.engagementRate || 0),
              source: user.source || 'dune',
              isActive: true,
              lastVerifiedAt: new Date(),
            }
          });
          savedCount++;
        }
      });
    });

    await clearCache();
    return savedCount;
  } catch (error) {
    console.error('[QualityUsersRepo] Error saving quality users:', error);
    throw error;
  }
}

export async function updateUserCoinStatus(
  fid: number,
  hasCoin: boolean,
  coinAddress?: string
): Promise<void> {
  try {
    await withRetry(() => prisma.qualityFarcasterUser.upsert({
      where: { fid },
      update: {
        hasCreatorCoin: hasCoin,
        creatorCoinAddress: coinAddress || null,
        lastCoinCheck: new Date(),
        updatedAt: new Date(),
      },
      create: {
        fid,
        hasCreatorCoin: hasCoin,
        creatorCoinAddress: coinAddress || null,
        lastCoinCheck: new Date(),
        isActive: true,
        source: 'coin_check',
      }
    }));
  } catch (error) {
    console.error(`[QualityUsersRepo] Error updating coin status for FID ${fid}:`, error);
  }
}

/**
 * Get profile for a specific FID
 */
export async function getProfileByFid(fid: number): Promise<QualityFarcasterUser | null> {
  try {
    const row = await prisma.qualityFarcasterUser.findUnique({
      where: { fid }
    });

    if (!row) return null;

    return {
      fid: row.fid,
      username: row.username || undefined,
      displayName: row.displayName || undefined,
      pfp: row.pfp || undefined,
      bio: row.bio || undefined,
      verifications: Array.isArray(row.verifications) ? row.verifications as any[] : [],
      url: (row as any).url || undefined,
      banner: (row as any).banner || undefined,
      primaryAddress: (row as any).primaryAddress || undefined,
      location: (row as any).location || undefined,
      followers: row.followers,
      following: row.following,
      totalCasts: row.totalCasts,
      engagementRate: Number(row.engagementRate),
      source: row.source || undefined,
      isActive: row.isActive,
      hasCreatorCoin: row.hasCreatorCoin,
      creatorCoinAddress: row.creatorCoinAddress || undefined,
      lastCoinCheck: row.lastCoinCheck || undefined,
    };
  } catch (error) {
    console.error(`[QualityUsersRepo] Error getting profile for FID ${fid}:`, error);
    return null;
  }
}

/**
 * Update user profile details
 */
export async function updateProfile(fid: number, data: Partial<QualityFarcasterUser>): Promise<void> {
  try {
    await withRetry(() => prisma.qualityFarcasterUser.upsert({
      where: { fid },
      update: {
        username: data.username,
        displayName: data.displayName,
        pfp: data.pfp,
        bio: data.bio,
        verifications: data.verifications,
        url: (data as any).url ?? undefined,
        banner: (data as any).banner ?? undefined,
        primaryAddress: (data as any).primaryAddress ?? undefined,
        location: (data as any).location ?? undefined,
        updatedAt: new Date(),
      },
      create: {
        fid,
        username: data.username || null,
        displayName: data.displayName || null,
        pfp: data.pfp || null,
        bio: data.bio || null,
        verifications: data.verifications || [],
        url: (data as any).url || null,
        banner: (data as any).banner || null,
        primaryAddress: (data as any).primaryAddress || null,
        location: (data as any).location || null,
        source: data.source || 'manual_sync',
        isActive: true,
      }
    }));
  } catch (error) {
    console.error(`[QualityUsersRepo] Error updating profile for FID ${fid}:`, error);
  }
}


/**
 * Seed initial quality users from hardcoded list
 */
export async function seedInitialQualityUsers(fids: number[]): Promise<number> {
  const users: QualityFarcasterUser[] = fids.map(fid => ({
    fid,
    source: 'manual',
    isActive: true,
  }));

  return saveQualityUsers(users);
}

export async function hasQualityUsers(): Promise<boolean> {
  try {
    const count = await prisma.qualityFarcasterUser.count({
      where: { isActive: true }
    });
    return count > 0;
  } catch (error) {
    console.error('[QualityUsersRepo] Error checking quality users:', error);
    return false;
  }
}

/**
 * Get statistics about quality users (follower distribution)
 */
export async function getQualityUsersStats() {
  try {
    const total = await prisma.qualityFarcasterUser.count({
      where: { isActive: true }
    });

    const over5k = await prisma.qualityFarcasterUser.count({
      where: {
        isActive: true,
        followers: { gte: 5000 }
      }
    });

    const over8k = await prisma.qualityFarcasterUser.count({
      where: {
        isActive: true,
        followers: { gte: 8000 }
      }
    });

    // We can use groupBy for the distribution, but since the ranges are custom, 
    // a small set of queries might be simpler or we use $queryRaw for efficiency if there are many.
    // Given the specific ranges in social.ts, let's use a single queryRaw for performance.
    const distributionResult: any[] = await prisma.$queryRaw`
      SELECT 
        CASE 
          WHEN followers >= 10000 THEN '10K+'
          WHEN followers >= 8000 THEN '8K-10K'
          WHEN followers >= 5000 THEN '5K-8K'
          WHEN followers >= 1000 THEN '1K-5K'
          WHEN followers >= 100 THEN '100-1K'
          WHEN followers >= 10 THEN '10-100'
          ELSE '<10'
        END as range,
        CAST(COUNT(*) AS INTEGER) as count
      FROM quality_farcaster_users
      WHERE is_active = TRUE
      GROUP BY range
    `;

    const rangeOrder: { [key: string]: number } = {
      '10K+': 1,
      '8K-10K': 2,
      '5K-8K': 3,
      '1K-5K': 4,
      '100-1K': 5,
      '10-100': 6,
      '<10': 7,
    };

    const distribution = distributionResult
      .map(row => ({
        range: row.range,
        count: row.count,
        percentage: total > 0 ? ((row.count / total) * 100).toFixed(1) : '0.0',
      }))
      .sort((a, b) => (rangeOrder[a.range] || 99) - (rangeOrder[b.range] || 99));

    return {
      total,
      over5k,
      over8k,
      percentages: {
        over5k: total > 0 ? ((over5k / total) * 100).toFixed(1) : '0.0',
        over8k: total > 0 ? ((over8k / total) * 100).toFixed(1) : '0.0',
      },
      distribution,
    };
  } catch (error) {
    console.error('[QualityUsersRepo] Error getting quality users stats:', error);
    throw error;
  }
}

/**
 * Clear the FIDs cache
 */
export async function clearCache(): Promise<void> {
  try {
    // Set with TTL of 1 second to expire immediately
    await set(CACHE_KEY, '[]', 1);
    console.log('[QualityUsersRepo] Cache cleared');
  } catch (error) {
    console.error('[QualityUsersRepo] Error clearing cache:', error);
  }
}

// Default export for compatibility
export default {
  getQualityFids,
  getQualityUsers,
  saveQualityUsers,
  updateUserCoinStatus,
  seedInitialQualityUsers,
  hasQualityUsers,
  getQualityUsersStats,
  clearCache,
  getProfileByFid,
  updateProfile,
};

File: kikoapi/internal/repositories/quality_users_repository.go
// Package repositories: Quality Farcaster users (from kiko-api repositories/qualityUsersRepository.ts).

package repositories

import (
	"context"

	"gorm.io/gorm"
)

// QualityFarcasterUser for get/save.
type QualityFarcasterUser struct {
	Fid                int
	Username           *string
	DisplayName        *string
	Pfp                *string
	Bio                *string
	Followers          *int
	Following          *int
	TotalCasts         *int
	EngagementRate     *float64
	Source             *string
	IsActive           bool
	HasCreatorCoin     bool
	CreatorCoinAddress *string
}

// QualityUsersRepository handles quality_farcaster_users table.
type QualityUsersRepository struct {
	db *gorm.DB
}

// NewQualityUsersRepository creates a QualityUsersRepository.
func NewQualityUsersRepository(database *gorm.DB) *QualityUsersRepository {
	return &QualityUsersRepository{db: database}
}

// GetQualityFids returns active FIDs ordered by followers desc, limit 2000.
func (r *QualityUsersRepository) GetQualityFids(ctx context.Context) ([]int, error) {
	var fids []int
	err := r.db.WithContext(ctx).Raw(
		`SELECT fid FROM quality_farcaster_users WHERE is_active = true ORDER BY followers DESC LIMIT 2000`,
	).Scan(&fids).Error
	return fids, err
}

// GetQualityUsers returns active users with limit.
func (r *QualityUsersRepository) GetQualityUsers(ctx context.Context, limit int) ([]QualityFarcasterUser, error) {
	if limit <= 0 {
		limit = 100
	}
	var rows []struct {
		Fid             int      `gorm:"column:fid"`
		Username        *string  `gorm:"column:username"`
		DisplayName     *string  `gorm:"column:display_name"`
		Pfp             *string  `gorm:"column:pfp"`
		Bio             *string  `gorm:"column:bio"`
		Followers       *int     `gorm:"column:followers"`
		Following       *int     `gorm:"column:following"`
		TotalCasts      *int     `gorm:"column:total_casts"`
		EngagementRate  *float64 `gorm:"column:engagement_rate"`
		Source          *string  `gorm:"column:source"`
		IsActive        bool     `gorm:"column:is_active"`
		HasCreatorCoin  bool     `gorm:"column:has_creator_coin"`
		CreatorCoinAddress *string `gorm:"column:creator_coin_address"`
	}
	err := r.db.WithContext(ctx).Raw(
		`SELECT fid, username, display_name, pfp, bio, followers, following, total_casts, engagement_rate, source, is_active, has_creator_coin, creator_coin_address FROM quality_farcaster_users WHERE is_active = true ORDER BY followers DESC LIMIT ?`,
		limit,
	).Scan(&rows).Error
	if err != nil {
		return nil, err
	}
	out := make([]QualityFarcasterUser, 0, len(rows))
	for _, row := range rows {
		out = append(out, QualityFarcasterUser{
			Fid: row.Fid, Username: row.Username, DisplayName: row.DisplayName, Pfp: row.Pfp, Bio: row.Bio,
			Followers: row.Followers, Following: row.Following, TotalCasts: row.TotalCasts, EngagementRate: row.EngagementRate,
			Source: row.Source, IsActive: row.IsActive, HasCreatorCoin: row.HasCreatorCoin, CreatorCoinAddress: row.CreatorCoinAddress,
		})
	}
	return out, nil
}

// SaveQualityUsers upserts users (one by one).
func (r *QualityUsersRepository) SaveQualityUsers(ctx context.Context, users []QualityFarcasterUser) (int, error) {
	saved := 0
	for _, u := range users {
		err := r.db.WithContext(ctx).Exec(
			`INSERT INTO quality_farcaster_users (fid, username, display_name, pfp, bio, followers, following, total_casts, engagement_rate, source, is_active, has_creator_coin, creator_coin_address, created_at, updated_at)
			 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())
			 ON CONFLICT (fid) DO UPDATE SET username = EXCLUDED.username, display_name = EXCLUDED.display_name, pfp = EXCLUDED.pfp, bio = EXCLUDED.bio, followers = EXCLUDED.followers, following = EXCLUDED.following, total_casts = EXCLUDED.total_casts, engagement_rate = EXCLUDED.engagement_rate, is_active = EXCLUDED.is_active, has_creator_coin = EXCLUDED.has_creator_coin, creator_coin_address = EXCLUDED.creator_coin_address, updated_at = NOW()`,
			u.Fid, u.Username, u.DisplayName, u.Pfp, u.Bio, u.Followers, u.Following, u.TotalCasts, u.EngagementRate, u.Source, u.IsActive, u.HasCreatorCoin, u.CreatorCoinAddress,
		).Error
		if err != nil {
			return saved, err
		}
		saved++
	}
	return saved, nil
}

// GetActiveFids is an alias for GetQualityFids (stub compatibility).
func (r *QualityUsersRepository) GetActiveFids(ctx context.Context) ([]int, error) {
	return r.GetQualityFids(ctx)
}
